pkg/core/xds/rules.go

```diff
@@ -156,33 +165,73 @@ func BuildRules(list []PolicyItem) Rules {
 		return rules[i].Subset.NumPositive() > rules[j].Subset.NumPositive()
 	})
 
-	return rules
+	return rules, nil
+}
+
+func merge(confs []proto.Message) (proto.Message, error) {
+	if len(confs) == 0 {
+		return nil, nil
+	}
+
+	resultBytes := []byte{}
+	for _, conf := range confs {
+		confBytes, err := json.Marshal(conf)
+		if err != nil {
+			return nil, err
+		}
+		if len(resultBytes) == 0 {
+			resultBytes = confBytes
+			continue
+		}
+		resultBytes, err = jsonpatch.MergePatch(resultBytes, confBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	result, err := newConf(reflect.TypeOf(confs[0]))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal(resultBytes, result); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
```

Comment:
I really feel we could get rid of this unmarshal/marshall by just leveraging reflection and walk through the type but I guess this can be an optimization for later. By this I mean that you should be able to walk the field of your struct and set the left value the right one.

Reply:
Yep, we probably can do that. But now we have a small advantage, instead of explaining how merging works, we can just point users to [RFC 7396](https://www.rfc-editor.org/rfc/rfc7396). I don't know if we will be happy with this algorithm in the future, but for now, [RFC 7396](https://www.rfc-editor.org/rfc/rfc7396) covers our needs.

Reply:
We can still implement something that follows the RFC 7396. Anyway probably an improvement for the future